ignore property errors when parsing fixure factories

bluetech opened the pull request:
Previously, properties such as in the added test case were triggered
during collection, possibly executing unintended code. Let's skip them
instead.

We should probably skip all custom descriptors, however I am not sure
how to do this cleanly, so for now just skip properties.

Fixes #2234.

The-Compiler commented:
I wonder if this should go to the `features` branch - while it can be seen as a bugfix, it also seems to me like there's a possibility of breaking something which worked before with it.

The-Compiler commented on _pytest/fixtures.py:
```diff
@@ -1068,6 +1068,9 @@ def parsefactories(self, node_or_obj, nodeid=NOTSET, unittest=False):
         autousenames = []
         for name in dir(holderobj):
+            # Skip properties before the following getattr executes them.
+            if isinstance(getattr(type(holderobj), name, None), property):
```
hmm, maybe this should use `safe_getattr` from `_pytest.compat`? Probably a getattr on a type will never raise something, but you never know with metaclasses and stuff I guess 😆

ionelmc replied:
It is technically possible to raise any exception from `__getattr__` implemented on metaclass. Eg:
```pycon
>>> getattr(fields.Fields.a, 'a', None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/fields/__init__.py", line 301, in __getattr__
    raise TypeError("Field %r is already specified as required." % name)
TypeError: Field 'a' is already specified as required.
```

ionelmc replied:
Perhaps a better way to look at the object is to iterate on `holderobj.__dict__` (getattr will involve the descriptor protocol, even if it's a class).

RonnyPfannschmidt replied:
the real problem is, that avoiding the descriptor protocol correctly is tricky

ionelmc replied:
Hehe. By far this is my „favourite” python quirk:
```pycon
>>> class Desc(object):
...  def __get__(self, name, owner):
...   print 'Surprise!', name, owner
...
>>> class Foo(object):
...  d = Desc()
...
>>> Foo.d
Surprise! None <class '__main__.Foo'>
```

bluetech replied:
I will switch from `getattr(type(holderobj), name, None)` to `type(holderobj).__dict__.get(name)`, which is hopefully safer. Here I assume that `type(X)` has a `__dict__` for all `X`. I think this is true - even if I try something like this:

```
>>> class FooMeta(type):
...    __slots__ = ()
... 
>>> class Foo(metaclass=FooMeta):
...   pass
```

it still has a `__dict__`:

```
>>> type(Foo())
<class '__main__.Foo'>
>>> type(Foo()).__dict__
mappingproxy({...stuff...})
```

bluetech replied:
(Indeed the third argument to `type(...)` must be a dict, `None` or other types are not accepted)

bluetech commented:
> I wonder if this should go to the features branch - while it can be seen as a bugfix, it also seems to me like there's a possibility of breaking something which worked before with it.

I agree there is some slight breakage potential here. If you want me to change the target branch, let me know (the maintainers can also edit the PR).

The-Compiler commented:
Let's wait what others think - no strong feelings either way, but history shows it's usually better to err on the side of caution :wink:

RonnyPfannschmidt commented on _pytest/fixtures.py:
```diff
@@ -1068,6 +1068,9 @@ def parsefactories(self, node_or_obj, nodeid=NOTSET, unittest=False):
         autousenames = []
         for name in dir(holderobj):
+            # Skip properties before the following getattr executes them.
+            if isinstance(type(holderobj).__dict__.get(name), property):
```
this really should use `safe_getattr` instead
else this protects only against one specific case of error while staying open for dozens of others

bluetech replied:
What sort of cases are you thinking of? As I understand it, this line cannot throw, since:

- `type(holderobj)` always has a `__dict__` for all `holderobj`, and that `__dict__` is indeed a `dict` (as I mentioned in a comment above).

- Accessing the `__dict__` and checking `isinstance` can't trigger any Python voodoo.

On the other hand, using `safe_getattr` will still trigger voodoo, though it will continue if it throws.

That said, having properties on the class *type* is really getting farfetched, so using `safe_getattr` (or just `getattr`) is fine by me as well. Just let me know if you still want me to change it.

RonnyPfannschmidt replied:
@bluetech basically any implementation of the descriptor protocol that does something at the class level can trigger

on the other hand i also think its fine to trigger those objects,  since well, python mantra is "consenting adults" - we dont need to protect people from their own traps too much

nicoddemus replied:
Just an idea: what if instead of explicitly skipping properties we just make sure we only accept function objects? After all fixtures can only be created using functions.

RonnyPfannschmidt replied:
@nicoddemus well, thats more strict in a good way ^^ i like the idea good thought :)

bluetech commented:
Thanks for the comment. I've updated it again to just replace the original `getattr` to `safe_getattr`. I rewrote the commit message, please have a look.

This patch fixes my immediate problem; hopefully someone will fix the problem properly, but doing so is a bit over my head as a passerby (haven't even had a chance to use pytest fixtures yet).

I also added another tiny change I noticed, you can drop that commit if you want.

RonnyPfannschmidt commented:
yup, the current code is fine to merge - additional enhancements can be done as part of a new pr

nicoddemus commented:
> yup, the current code is fine to merge - additional enhancements can be done as part of a new pr

Agreed, thanks again @bluetech!
